File: tools/collector/piscapisca/normalize.py
# normalize.py — normalizadores e schema-alvo COMUM, ESPELHO fiel do lib/normalize.mjs (Node).
#
# PORQUÊ espelhado (e não reutilizado): este coletor é Python (o Scrapling é Python), mas tem de
# produzir EXATAMENTE o mesmo registo normalizado dos 23 coletores Node, para que os registos se
# fundam num único stream de upsert. Por isso replicamos aqui, byte-a-byte no comportamento, os
# CAMPOS_BASE e os normalizadores genéricos. Cada site tem depois o seu mapeamento (schema.py).

import re

# Campos-base comuns (ordem canónica — IDÊNTICA ao lib/normalize.mjs). Cada site acrescenta extras.
CAMPOS_BASE = [
    "make", "model", "variant", "year", "km", "fuel", "gearbox", "engine",
    "color", "doors", "category", "price", "currency", "country", "region",
    "postalCode", "source", "detail_url", "image", "collected_at",
]


def to_int(v):
    """Extrai o inteiro de uma string com ruído ("182 350 km", "€ 20.950", "170 cv").
    Devolve None se não houver dígitos (preferimos None a 0 para não falsear estatísticas).
    Espelha o toInt do lib/normalize.mjs (re.sub(r'\\D','') → int)."""
    if v is None:
        return None
    digits = re.sub(r"\D", "", str(v))
    return int(digits) if digits else None


def clean_str(v):
    """trim + colapso de espaços/quebras internas. Devolve None se ficar vazio.
    Espelha o cleanStr do lib/normalize.mjs."""
    if v is None:
        return None
    s = re.sub(r"\s+", " ", str(v)).strip()
    return s or None

File: tools/collector/lib/normalize.ts
// lib/normalize.ts — normalizadores e schema-alvo COMUM a todos os coletores.
//
// PORQUÊ partilhado: todos os coletores (theparking, autotrader, …) produzem o mesmo
// registo normalizado, para comparar preços PT vs. UE de forma uniforme. Aqui ficam os
// campos-base e os normalizadores genéricos; cada site tem o seu próprio mapeamento
// (schema.ts) que usa estes helpers.

// Campos-base comuns (ordem canónica). Cada site pode acrescentar extras próprios.
export const CAMPOS_BASE = [
  'make', 'model', 'variant', 'year', 'km', 'fuel', 'gearbox', 'engine',
  'color', 'doors', 'category', 'price', 'currency', 'country', 'region',
  'postalCode', 'source', 'detail_url', 'image', 'collected_at',
] as const;

// CollectorRecord — o registo normalizado COMUM a todos os coletores (os 20 CAMPOS_BASE).
// É o contrato verificável pelo compilador: cada coletor define o seu `<X>Record extends
// CollectorRecord` com os extras próprios do site. Alinha com lib/types.ts do frontend.
export interface CollectorRecord {
  make: string | null;
  model: string | null;
  variant: string | null;
  year: number | null;
  km: number | null;
  fuel: string | null;
  gearbox: string | null;
  engine: string | number | null;
  color: string | null;
  doors: number | null;
  category: string | null;
  price: number | null;
  currency: string;
  country: string | null;
  region: string | null;
  postalCode: string | null;
  source: string | null;
  detail_url: string | null;
  image: string | null;
  collected_at: string | null;
}

// toInt: extrai o inteiro de uma string com ruído ("217.828 km", "€ 1.975", "5 Doors").
// Devolve null se não houver dígitos (preferimos null a 0 para não falsear estatísticas).
export function toInt(v: unknown): number | null {
  if (v == null) return null;
  const digits = String(v).replace(/[^\d]/g, '');
  return digits ? Number(digits) : null;
}

// cleanStr: trim + colapso de espaços/quebras internas. Devolve null se ficar vazio.
export function cleanStr(v: unknown): string | null {
  if (v == null) return null;
  const s = String(v).replace(/\s+/g, ' ').trim();
  return s || null;
}
